[clang][TSA] Make RequiresCapability a DeclOrType attribute

tbaederr opened the pull request:
This was much easier than expected actually.

I'm sure I'm missing a lot of changes and test coverage though.

@aaronpuchert I'm starting a new review here instead of continuing the one in Phab since this is a completely different attempt. I hope that's ok.

tbaederr committed:
[clang][TSA] Make RequiresCapability a DeclOrType attribute

tbaederr commented on clang/test/Sema/warn-thread-safety-analysis.c:
```diff
@@ -136,6 +136,17 @@ int main(void) {
 
+  /// Function pointers
+  {
+    int __attribute__((requires_capability(&mu1))) (*function_ptr)(int) = Foo_fun1;
```
This assignment also works if `Foo_fun1` does not have the `requires_capability(&mu1)` attribute, it would be better if we warned here I guess.

AaronBallman replied:
Er, not certain I agree that should be a separate change -- I don't see *anything* hooking this up as a type attribute beyond the change to `DeclOrTypeAttr` in `Attr.td`. Declaration attributes have a ton of tablegenerated code that hooks things up for them, but type attributes are special and require more involved manual work.

I would have expected to see changes [in here](https://github.com/llvm/llvm-project/blob/28b7e281d4eaea0d5d56b1a4cf7a550be746a007/clang/lib/Sema/SemaType.cpp#L8615) to hook this up with the type system. Once the type is wrapped in an attribute, the diagnostic behavior you're looking for should come along "for free".

tbaederr commented:
Ping

aaronpuchert reviewed (commented):
> This was much easier than expected actually.

Making it a `DeclOrType` attribute is indeed a nice idea, this would allow existing attributes to stay where they are. Is it still inheritable, i.e. does it also apply to later redeclarations?

Of course I'm also wondering why we don't have to change anything in the analysis, aren't we currently only looking at declaration attributes, and wouldn't consider attributes on the type? Or is there some mechanism that gives us the combined set of declaration and type attributes?

> I'm starting a new review here instead of continuing the one in Phab since this is a completely different attempt. I hope that's ok.

Yeah, since it's a completely different approach than https://reviews.llvm.org/D152246 we're not really losing any history.

tbaederr commented:
> > This was much easier than expected actually.
> 
> Making it a `DeclOrType` attribute is indeed a nice idea, this would allow existing attributes to stay where they are. Is it still inheritable, i.e. does it also apply to later redeclarations?

I haven't tested this explicitly, but `DeclOrTypeAttr` is-a `InheritableAttr`:
https://github.com/llvm/llvm-project/blob/5dd9266c4e0b0a02cbb02a29bf2606224ccf2820/clang/include/clang/Basic/Attr.td#L644-L647C1

> Of course I'm also wondering why we don't have to change anything in the analysis, aren't we currently only looking at declaration attributes, and wouldn't consider attributes on the type? Or is there some mechanism that gives us the combined set of declaration and type attributes?

Let me check why this just works.

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3315,7 +3315,7 @@ def RequiresCapability : InheritableAttr {
   let ParseArgumentsAsUnevaluated = 1;
   let InheritEvenIfAlreadyPresent = 1;
-  let Subjects = SubjectList<[Function]>;
+  /*let Subjects = SubjectList<[Function]>;*/
```
Instead of removing it, can this be updated to be correct (we have `FunctionPointer` as a subject)?

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -8141,6 +8141,16 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
 
+  // We allow this on function declaration as well as
+  // variable declarations of function pointer type.
+  if (!D->isFunctionPointerType() && !isa<FunctionDecl>(D)) {
```
Once the Subjects list is corrected, I believe this code can be removed.

AaronBallman reviewed (commented):
The changes also need a release note at some point.

tbaederr commented:
Ping

AaronBallman commented:
> Ping

Did you see https://github.com/llvm/llvm-project/pull/67095#discussion_r1355468253 ?

tbaederr commented:
> > Ping
> 
> Did you see [#67095 (comment)](https://github.com/llvm/llvm-project/pull/67095#discussion_r1355468253) ?

Yes, I added the changes to `processTypeAttrs()` and added the subject list back. Is there something I'm missing?

But the tests fail too, I'll check that locally.

AaronBallman commented:
> > > Ping
> > 
> > 
> > Did you see [#67095 (comment)](https://github.com/llvm/llvm-project/pull/67095#discussion_r1355468253) ?
> 
> Yes, I added the changes to `processTypeAttrs()` and added the subject list back. Is there something I'm missing?

Nope, this was another lovely case of GitHub hiding details from me. The Files tab wasn't showing me your changes to that file for me... until I clicked it a second time and then new changes showed up. Sorry for the noise!

AaronBallman commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8602,6 +8602,31 @@ static void HandleAnnotateTypeAttr(TypeProcessingState &State,
+    return;
+  }
+
+  // FIXME: We need to sanity check the arguments here I think? Like we do in
+  // SemaDeclAtr.cpp.
```
Yup, this needs to be done manually because the helper code automatically generated by tablegen isn't yet smart enough to generate logic for type attributes. I think this work should be done as part of this commit instead of left as follow-up work.

AaronBallman commented on clang/test/Sema/warn-thread-safety-analysis.c:
```diff
@@ -136,6 +136,17 @@ int main(void) {
+    int five = function_ptr(5);
+    mutex_exclusive_unlock(&mu1);
+  }
+
```
We'll also need changes showing that we properly diagnose all the usual stuff from the decl form of the attribute but with the type form instead, plus additional coverage for new diagnostics like conversions between function pointer types with/out the attribute.

tbaederr replied:
There are no such diagnostics when converting between function pointers. Are there supposed to be any, without adding special handling for this? I quickly checked and `noreturn` has the same behavior.

tbaederr commented:
It's pretty easy to get an error like:
```
./tsa.c:77:18: error: incompatible function pointer types assigning to 'int (*)(int) __attribute__((requires_capability(blah)))' from 'int (void)' [-Wincompatible-function-pointer-types]
   77 |     function_ptr = foo2;
      |                  ^ ~~~~
```

but as soon as I change `foo2` to also take an int parameter, the error vanishes completely.

AaronBallman commented on clang/lib/AST/TypePrinter.cpp:
```diff
@@ -1894,6 +1894,8 @@ void TypePrinter::printAttributedAfter(const AttributedType *T,
     OS << "__clang_arm_mve_strict_polymorphism";
     break;
+  case attr::RequiresCapability:
+    OS << "requires_capability(blah)";
```
`blah`?

AaronBallman replied:
Not many type attributes have arguments to them -- but `pcs` does print its argument but `annotate_type` doesn't. If you've got access to the information, best to print it, but if you don't have what you need to print the arguments, then I'd go with `(...)` as the argument list with a FIXME comment.

erichkeane replied:
See aaron's comment above.

AaronBallman commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8938,6 +8957,11 @@ static void processTypeAttrs(TypeProcessingState &state, QualType &type,
       attr.setUsedAsTypeAttr();
       break;
     }
+    case ParsedAttr::AT_RequiresCapability: {
+      HandleRequiresCapabilityAttr(state, type, attr);
+      attr.setUsedAsTypeAttr();
+      break;
+    }
```
Okay, maybe we do want to do this change separately. I think I was wrong about modeling after this; I think we want it to be done like `handleFunctionTypeAttr()` and `FUNCTION_TYPE_ATTRS_CASELIST` so that it becomes part of the `FunctionType` itself. That's what would get us the diagnostics on type mismatches (we look through `AttributedType` as sugar: https://github.com/llvm/llvm-project/blob/28b7e281d4eaea0d5d56b1a4cf7a550be746a007/clang/include/clang/AST/Type.h#L4999)

AaronBallman replied:
Hmmmm, my intuition is that if we don't do this work now, we'll forever be stuck with the current behavior where we don't diagnose pointer mismatches because someone will rely on that continuing to work. So maybe we do want to do this work now...

CC @erichkeane @aaronpuchert for opinions on whether we think this needs to diagnose mismatches in the initial offering of it as a type attribute.

erichkeane replied:
I think I agree Aaron, if we don't diagnose now, it'll be pretty ugly to do it later.

tbaederr replied:
Meh, it seems like adding a pointer to `FunctionType::ExtInfo` would blow it up a lot, and just a bit isn't enough to save the information.

erichkeane replied:
There were a few changes there recently... i wonder if it is time to just accept that we're going all-in on trailing-storage for this type.  We're now doing it for the ARM SVE types, plus the 'lesser used' structure, so perhaps the 'capability pointer' can end up there too.  WDYT?

tbaederr replied:
@AaronBallman Does this sound like the way to go to you as well?

AaronBallman replied:
No, we don't want to modify `FunctionType::ExtInfo`, I think what @erichkeane was talking about is that we added trailing objects to `FunctionProtoType` that we can squirrel lesser-used data into. That might be reasonable here because I cannot imagine anyone using thread safety attributes with K&R C interfaces, but that would be the downside to putting the data on `FunctionProtoType`.

erichkeane replied:
It would kind of stink to have to do it in 2 places, but it could go in trailing-storage/etc on `FunctionProtoType` and `FunctionNoProtoType`, then just disambiguate it with a function.  So it doesn't need to end up in `FunctionType::ExtInfo` or `FunctionType` itself.

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3341,7 +3341,7 @@ def RequiresCapability : InheritableAttr {
+  let Subjects = SubjectList<[Function, FunctionPointer]>;
   let Accessors = [Accessor<"isShared", [Clang<"requires_shared_capability", 0>,
                                          Clang<"shared_locks_required", 0>]>];
   let Documentation = [Undocumented];
```
I think now is a good time to insist we document this :)  It would at least make review easier...

AaronBallman replied:
FWIW: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

So perhaps we could give some very high-level documentation but link to there? (At the very least, hopefully this helps you with review efforts.)

erichkeane replied:
Oooh, yeah, i think that would be enough if this/others linked to it.  I went to our attributes page to figure out what this does, and saw "not documented".  Just a 2-sentence summary + "This is part of Thread Safety Analysis, as documented here" would be great!  

Also, jsut figured out what "TSA" refers to here...

tbaederr commented:
Ping

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3341,7 +3341,7 @@ def RequiresCapability : InheritableAttr {
+  let Subjects = SubjectList<[Function, FunctionPointer]>;
   let Accessors = [Accessor<"isShared", [Clang<"requires_shared_capability", 0>,
                                          Clang<"shared_locks_required", 0>]>];
   let Documentation = [Undocumented];
```
Still isn't documented, see discussion with Aaron above.

erichkeane reviewed (commented):
Seems you pinged this without the commit to fix the things we've requested?

tbaederr commented:
> Seems you pinged this without the commit to fix the things we've requested?

Right, I was more looking for guidance what to do with  `FunctionType::ExtInfo`. It seems like adding the requires capability to that would be needed to make it a function type attribute, `ExtInfo` currently only saves a few bits and we need a full pointer.

tbaederr commented:
One thing I'm unsure about right now:  This probably needs more work, especially regarding assigning function pointers, doesn't it? Even in the test case I added right now, `Foo_fun2` is declared with `SHARED_LOCKS_REQUIRED(mu2) EXCLUSIVE_LOCKS_REQUIRED(mu1)`, but then the function pointer I assign it to only requires `mu1` - so calling it via that function pointer makes it possible to call the function without holding `mu2`. 
Assigning it to the function pointer is only OK if the function pointer type needs a superset (or the same set) of the mutexes the function needs, otherwise it would ideally be an error. But not sure if we currently do anything similar for other TSA stuff.
